src/Compilers/Core/Portable/Serialization/StreamObjectWriter.cs

```diff
@@ -983,60 +707,49 @@ private void WriteObject(object instance)
             }
             else
             {
-                Action<ObjectWriter, object> typeWriter;
-                if (!_binder.TryGetWriter(instance, out typeWriter))
+                var writable = instance as IObjectWritable;
+                if (writable == null)
                 {
-                    throw NoSerializationWriterException(instance.GetType().FullName);
+                    throw NoSerializationWriterException($"{instance.GetType()} must implement {nameof(IObjectWritable)}");
                 }
 
-                if (_recursive)
-                {
-                    _recursionDepth++;
-                    StackGuard.EnsureSufficientExecutionStack(_recursionDepth);
-                    if (_recursionDepth > MaxRecursionDepth)
-                    {
-                        throw new RecursionDepthExceeded();
-                    }
-
-                    // emit object header up front
-                    this.WriteObjectHeader(instance, 0);
-
-                    typeWriter(this, instance);
+                var oldDepth = _recursionDepth;
+                _recursionDepth++;
 
-                    _recursionDepth--;
+                if (_recursionDepth % MaxRecursionDepth == 0)
+                {
+                    // If we're recursing too deep, move the work to another thread to do so we
+                    // don't blow the stack.
+                    var task = Task.Factory.StartNew(
+                        () => WriteObjectWorker(instance, writable), 
+                        _cancellationToken,
+                        TaskCreationOptions.LongRunning,
+                        TaskScheduler.Default);
```

Comment:
We're nto using the threadpool here.  We're literally creating our own thread.  That's what LongRunning is for.  I don't want to starve the threadpool.